datafusion/expr/src/logical_plan/invariants.rs

```diff
@@ -41,19 +61,54 @@ pub enum InvariantLevel {
     Executable,
 }
 
+/// Apply the [`InvariantLevel::Always`] check at the root plan node only.
 pub fn assert_always_invariants(plan: &LogicalPlan) -> Result<()> {
     // Refer to <https://datafusion.apache.org/contributor-guide/specification/invariants.html#relation-name-tuples-in-logical-fields-and-logical-columns-are-unique>
     assert_unique_field_names(plan)?;
 
     Ok(())
 }
 
+/// Visit the plan nodes, and confirm the [`InvariantLevel::Executable`]
+/// as well as the less stringent [`InvariantLevel::Always`] checks.
 pub fn assert_executable_invariants(plan: &LogicalPlan) -> Result<()> {
+    // Always invariants
     assert_always_invariants(plan)?;
+    assert_valid_extension_nodes(plan, InvariantLevel::Always)?;
+
+    // Executable invariants
+    assert_valid_extension_nodes(plan, InvariantLevel::Executable)?;
```

Comment:
Presuming that the extension nodes are not at the plan root, these are checked only during the recursive assertion (a.k.a. `assert_executable_invariants`). Here you see that the always and executable invariants, for the extension nodes, are both checked.